### PR TITLE
Make both `torch.amp` and `apex.amp` available as backend for mixed precision training

### PR DESCRIPTION
**Pull request by NaleRaphael**
Hi @davidtvs, here is a summary of changes made in this PR for issue #67, #90.

## Required changes to integrate `torch.amp`
- `torch.cuda.amp.GradScaler` has to be passed into `LRFinder`, because it's will be used in the following stages:
  - backward pass:
    ```python
    # without AMP (automatic precision training), or using `apex.amp`
    loss.backward()
    
    # using `torch.amp`
    scaler.scale(loss).backward()
    ```
  - optimizer steps:
    ```python
    # without AMP, or using `apex.amp`
    optimizer.step()
    
    # using `torch.amp`
    scaler.step(optimizer)
    scaler.update()
    ```
- `torch.amp.autocast()` need to be called in forward pass:
  ```python
  # without AMP, or using `apex.amp`
  outputs = self.model(inputs)
  loss = self.criterion(outputs, labels)
  
  # using `torch.amp`
  with torch.amp.autocast(device_type=..., dtype=...):
    outputs = self.model(inputs)
    loss = self.criterion(outputs, labels)
  ```

## Proposed changes
3 new keyword arguments are added to `LRFinder.__init__()`:
- `amp_backend`: a string to select AMP backend
- `amp_config`: a dict to store arguments required by `torch.amp.autocast()`
- `grad_scaler`: a `torch.cuda.amp.GradScaler` instance to be used in `LRFinder._train_batch()`

This should maximize the flexibility for user to control how AMP works with `LRFinder`.

If there is a need to apply advanced tricks with `torch.amp` (e.g., for multi-GPUs/models/losses [1][1]), it's still achievable by just overriding `LRFinder._train_batch()`. So we can focus on the current implementation for gradient accumulation without worrying about other variants.

## Note
The new script `examples/mnist_with_amp.py` can be used to check the results produced by running `LRFinder` with different AMP backends. Here are the results produced on my machine with command `$ python mnist_with_amp.py --batch_size=32 --tqdm --amp_backend=...`:

|![][mnist_amp_none]|![][mnist_amp_apex]|![][mnist_amp_torch]|
|:-:|:-:|:-:|
|without AMP|`apex.amp`|`torch.amp`|

\* In these 3 figures, suggested LRs are all the same one: 2.42E-01

Package information:
- python: 3.9.18
- torch: 1.13.1+cu117
- apex: 0.1 (compiled from source, and I need to checkout to revision `2386a912164b0c5cfcd8be7a2b890fbac5607c82` to build. see also [this comment][apex_build_issue])

As always, feel free to let me know if there is anything can be improved.

[1]: https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers
[mnist_amp_none]: https://github.com/davidtvs/pytorch-lr-finder/assets/7083478/3a17c618-2490-4ece-9efa-dc12523a9470
[mnist_amp_apex]: https://github.com/davidtvs/pytorch-lr-finder/assets/7083478/f02de233-8a2c-4c97-bcd9-3afe1293afba
[mnist_amp_torch]: https://github.com/davidtvs/pytorch-lr-finder/assets/7083478/1f36e205-c83e-4027-bb61-5d3137ab6e91
[apex_build_issue]: https://github.com/NVIDIA/apex/issues/1735#issuecomment-1743179030

### CODE REVIEW

**Comment by NaleRaphael** [comment]
According to this issue https://github.com/actions/setup-python/issues/544, it seems supports for Python<3.6 are dropped after the base image is updated to Ubuntu 22.04.
Also, `torch.amp` is released in [PyTorch 1.6][1], and the newly added unit tests are designed to test it to run with GPU. So it might not be able to test with current CI setup.

[1]: https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/

**Comment by davidtvs** [comment]
@NaleRaphael, thanks for the PR, looks good. I've updated the CI workflow, it should be functional again. You probably have to rebase this branch.

**Comment by NaleRaphael** [comment]
Sure, wait for a moment and I'll update it.

**Comment by NaleRaphael** [comment]
Fixed f-string issue (to be compatible with Py 3.5).

By the way, there is one thing I forget to mention. Since the default value of the new keyword argument `amp_backend` is `None`, AMP won't be enabled even though `apex` has been installed. The behavior is different to previous revisions because it will use `apex.amp` automatically if `apex` is available.

This might be a breaking change to user. However, current behavior seems more reasonable to me now. Because mixed precision training provided by `apex` would only work with GPUs with tensor cores (RTX series). Though it can run on some pre-Turing cards, performance won't be as good as training in fp32, see also [this comment][1]. Therefore, defaulting `amp_backend` to `None` should avoid resulting some surprises to users.

[1]: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

**Review by davidtvs — APPROVED** [review]
LG! Thanks @NaleRaphael
